src-java/stats-topology/stats-storm-topology/src/test/java/org/openkilda/wfm/topology/stats/StatsTopologyTest.java

```diff
@@ -845,212 +1639,326 @@ private void validateFlowStats(
+                datapointMap.get(METRIC_PREFIX + "haflow.meter.bits").get(1).getValue().longValue());
+        assertEquals(meterStats.getByteInCount(),
+                datapointMap.get(METRIC_PREFIX + "haflow.meter.bytes").get(1).getValue().longValue());
+        assertEquals(meterStats.getPacketsInCount(),
+                datapointMap.get(METRIC_PREFIX + "haflow.meter.packets").get(1).getValue().longValue());
 
-        Assert.assertEquals(3, rawPacketsMetric.size());
-        for (Datapoint entry : rawPacketsMetric) {
-            Map<String, String> tags = entry.getTags();
-            Assert.assertEquals(CookieType.SERVER_42_FLOW_RTT_INGRESS.name().toLowerCase(), tags.get("type"));
 
-            if (Objects.equals(0, entry.getValue())) {
-                Assert.assertEquals("unknown", tags.get("flowid"));
-            } else if (Objects.equals(1, entry.getValue())) {
-                Assert.assertEquals(flowId, tags.get("flowid"));
-            } else if (Objects.equals(2, entry.getValue())) {
-                Assert.assertEquals("unknown", tags.get("flowid"));
-            } else {
-                Assert.fail(format("Unexpected metric value: %s", entry));
+        String direction = cookie.getDirection().name().toLowerCase();
+        Map<String, List<Datapoint>> datapointMapOfLists = createDatapointMapOfLists(datapoints);
+        datapointMapOfLists.keySet().forEach(metricName -> {
+            switch (metricName) {
+                case METRIC_PREFIX + "haflow.meter.ypoint.bits":
+                case METRIC_PREFIX + "haflow.meter.ypoint.bytes":
+                case METRIC_PREFIX + "haflow.meter.ypoint.packets":
+                    Datapoint datapoint = datapointMapOfLists.get(metricName).get(0);
+                    assertDataPoint(haFlowId, datapoint, false, 3, switchId, meterId,
+                            null, null, null);
+                    assertDataPoint(haFlowId, datapoint, false, 3, switchId, meterId,
+                            null, null, null);
+                    assertDataPoint(haFlowId, datapoint, false, 3, switchId, meterId,
+                            null, null, null);
+                    break;
+                case METRIC_PREFIX + "haflow.meter.bits":
+                case METRIC_PREFIX + "haflow.meter.bytes":
+                case METRIC_PREFIX + "haflow.meter.packets":
+                    List<Datapoint> currentDatapoints = datapointMapOfLists.get(metricName)
+                            .stream().sorted((o1, o2) -> {
+                                if (Objects.equals(o1, o2)) {
+                                    return 0;
+                                }
+                                if (o1.hashCode() > o2.hashCode()) {
+                                    return 1;
+                                } else {
+                                    return -1;
+                                }
+                            }).collect(toList());
```

Comment:
It is vary unstable solution. When `Datapoint` will be changed its hashcode will be changed. It means datapoins can be sorted in the difference way. 

I'm suggesting this solution

```
                case METRIC_PREFIX + "haflow.meter.bits":
                case METRIC_PREFIX + "haflow.meter.bytes":
                case METRIC_PREFIX + "haflow.meter.packets":
                    Map<String, Datapoint> dataPointsByFlowId = datapointMap.get(metricName).stream()
                            .collect(toMap(metric -> metric.getTags().get("flowid"), Function.identity()));
                    Map<String, FlowSegmentCookie> cookieMap = new HashMap<>();
                    cookieMap.put(subFlowId1, cookieSubFlow1);
                    cookieMap.put(subFlowId2, cookieSubFlow2);

                    for (String subflowId : new String[] {subFlowId1, subFlowId2}) {
                        Datapoint subFlowDataPoint = dataPointsByFlowId.get(subflowId);
                        assertNotNull(subFlowDataPoint);
                        assertEquals(6, subFlowDataPoint.getTags().size());

                        assertEquals(switchId.toOtsdFormat(), subFlowDataPoint.getTags().get("switchid"));
                        assertEquals(Long.toString(meterId), subFlowDataPoint.getTags().get("meterid"));
                        assertEquals(haFlowId, subFlowDataPoint.getTags().get("ha_flow_id"));
                        assertEquals(subflowId, subFlowDataPoint.getTags().get("flowid"));
                        assertEquals(direction, subFlowDataPoint.getTags().get("direction"));
                        assertEquals(Long.toString(cookieMap.get(subflowId).getValue()),
                                subFlowDataPoint.getTags().get("cookie"));
                    }
                    break;
```

Reply:
good solution, thanks